### PR TITLE
Rethrow BatchParserExceptions, swallow the stack trace and add the name of the file in error

### LINKED ISSUE #128: Can not recursively exclude files from build
Hello, 

It looks like it's impossible to add entire directory to exclude from build. Whenever I'm trying to exclude a folder containing more than one file with command `<Content Remove="Repos\*\Scripts\Inserts\*.sql" />`, it throws an error:

> ---> Microsoft.SqlTools.ServiceLayer.BatchParser.BatchParserException: Incorrect syntax was encountered while parsing '
>   '.
>      at Microsoft.SqlTools.ServiceLayer.BatchParser.Parser.RaiseError(ErrorCode errorCode, Token token, String message) in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/sqltoolsservice/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/Parser.cs:line 621
>      at Microsoft.SqlTools.ServiceLayer.BatchParser.Parser.RaiseError(ErrorCode errorCode, String message) in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/sqltoolsservice/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/Parser.cs:line 612
>      at Microsoft.SqlTools.ServiceLayer.BatchParser.Parser.Expect(LexerTokenType lexerTokenType) in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/sqltoolsservice/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/Parser.cs:line 217
>      at Microsoft.SqlTools.ServiceLayer.BatchParser.Parser.ParseInclude() in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/sqltoolsservice/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/Parser.cs:line 272
>      at Microsoft.SqlTools.ServiceLayer.BatchParser.Parser.ParseLines() in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/sqltoolsservice/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/Parser.cs:line 394
>      at Microsoft.SqlTools.ServiceLayer.BatchParser.Parser.Parse() in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/sqltoolsservice/src/Microsoft.SqlTools.ManagedBatchParser/BatchParser/Parser.cs:line 232
>      at MSBuild.Sdk.SqlProj.DacpacTool.ScriptParser.Parse() in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/src/DacpacTool/ScriptParser.cs:line 
> 27
>      at MSBuild.Sdk.SqlProj.DacpacTool.ScriptParser.CollectFileNames() in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/src/DacpacTool/ScriptParser.cs:line 34
>      at MSBuild.Sdk.SqlProj.DacpacTool.ScriptInspector.AddIncludedFiles(FileInfo file) in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/src/DacpacTool/ScriptInspector.cs:line 28
>      at MSBuild.Sdk.SqlProj.DacpacTool.Program.InspectIncludes(InspectOptions options) in /home/runner/work/MSBuild.Sdk.SqlProj/MSBuild.Sdk.SqlProj/src/DacpacTool/Program.cs:line 149
>      --- End of inner exception stack trace ---
>      at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
>      at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
>      at System.Delegate.DynamicInvokeImpl(Object[] args)
>      at System.Delegate.DynamicInvoke(Object[] args)
>      at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
>      at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseErrorReporting>b__21_0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseVersionOption>b__0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseTypoCorrections>b__0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__22_0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseParseDirective>b__20_0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseDebugDirective>b__11_0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
>   --- End of stack trace from previous location ---
>      at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()

### PR DESCRIPTION
**Pull request by jeffrosenberg**
Fixes #128

### CODE REVIEW

**Comment by jeffrosenberg** [comment]
This is my attempt to throw a more user-friendly error when we get a BatchParserException, suggestions on how to improve it are welcome 😃

**Review by ErikEJ — APPROVED** [review]

File: src/DacpacTool/ScriptParser.cs [hunk 1]
```diff
@@ -24,8 +25,20 @@ private void Parse()
...
             if (! _parsed)
             {
-                _parser.Parse();
-                _parsed = true;
+                try
+                {
+                    _parser.Parse();
+                    _parsed = true;
+                }
+                catch (Microsoft.SqlTools.ServiceLayer.BatchParser.BatchParserException ex)
```
Comment (jmezach) [inline src/DacpacTool/ScriptParser.cs:33]: Do we need to fully qualify the exception type? There's already a using for this namespace in this file?

File: src/DacpacTool/ScriptParser.cs [hunk 2]
```diff
@@ -24,8 +25,20 @@ private void Parse()
...
+                {
+                    string mostRecentFile = _includedFileNames.LastOrDefault();
+                    if (mostRecentFile != null)
```
Comment (jmezach) [inline src/DacpacTool/ScriptParser.cs:36]: I would prefer not to have two different throw statements, but put the exception message into a variable (possibly using a ternary operator) and then a single throw statement.

File: test/DacpacTool.Tests/ScriptParserTests.cs [hunk 3]
```diff
@@ -162,5 +162,16 @@ public void CanGenerateScriptWithNestedIncludesContainingRelativePaths()
...
+
+        [TestMethod]
+        public void ParserFailsWhenSyntaxError()
+        {
+            // Arrange
+            var collector = new ScriptParser($"{TEST_FOLDER}/TestProjectWithIncludedSyntaxErrors/Post-Deployment/Script.PostDeployment.sql", _variableResolver);
+
+            // Act / Assert
+            Should.Throw<System.InvalidOperationException>(() => collector.CollectFileNames())
+                .Message.ShouldBe("Incorrect syntax was encountered while parsing '\n'. File: DataScript.sql");
```
Comment (jmezach) [inline test/DacpacTool.Tests/ScriptParserTests.cs:174]: The message here feels a bit weird to me, especially with the `\n` before the dot. Is that deliberate?

File: test/TestProjectWithIncludedSyntaxErrors/Post-Deployment/DataScript.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+print 'Loading file '
```
Comment (jmezach) [inline test/TestProjectWithIncludedSyntaxErrors/Post-Deployment/DataScript.sql:1]: Do we need all these test files just to trigger the error? Feels like a contrived example scenario that could be simplified for easier maintainability.

File: test/TestProjectWithIncludedSyntaxErrors/TestProjectWithIncludedSyntaxErrors.csproj [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<Project Sdk="MSBuild.Sdk.SqlProj/1.16.0-beta.4">
```
Comment (jmezach) [inline test/TestProjectWithIncludedSyntaxErrors/TestProjectWithIncludedSyntaxErrors.csproj:1]: Perhaps we could setup this project to use the `Sdk.props` and `Sdk.targets` files directly as the other test projects do?

**Review by jmezach — CHANGES_REQUESTED** [review]
@jeffrosenberg Thanks for this! Have a couple of comments.

File: src/DacpacTool/ScriptParser.cs (hunk 2 above)
Comment (jeffrosenberg) [reply src/DacpacTool/ScriptParser.cs:36]: heh, that was my original approach but I decided I didn't like it. 😄 I don't have a strong opinion, though, just thought it looked a little messy with the ternary. I'll go back to this approach with the next commit.

File: test/DacpacTool.Tests/ScriptParserTests.cs (hunk 3 above)
Comment (jeffrosenberg) [reply test/DacpacTool.Tests/ScriptParserTests.cs:174]: The first half of the error comes from the test scenario I imported, per my previous comment, I just matched exactly what was coming back from the parser error.

**Comment by jeffrosenberg** [comment]
@jmezach I've addressed most of your comments, except for the two related to the new test project; I didn't have time to work on that tonight. Please take a look at my comments and if you'd like me to give that another shot, I can definitely do that, but it's a holiday weekend here in the US so it might be a few days before I get that pushed. 🙂

**Comment by jmezach** [comment]
Changes look good to me. I do think we should try to create the simplest test scenario that we possibly can, to avoid shooting ourselves in the foot later on, but obviously there's no need to rush this :). Enjoy the holiday weekend! I'll probably be a little bit less responsive in the coming weeks as well, since I got the keys to my new house last Thursday so there's a lot of things that will need to happen now.

**Comment by jeffrosenberg** [comment]
Just checking in on this to let you know that I'll have to finish it up next week. I didn't get to this over the last couple of days and now I'm heading out for a few days of vacation. Not much left here, I'll definitely wrap it up next week 🙂 

> I'll probably be a little bit less responsive in the coming weeks as well, since I got the keys to my new house last Thursday

Congrats, @jmezach, enjoy!

**Comment by jeffrosenberg** [comment]
@jmezach I've simplified the test scenarios. In doing so, I've realized that the scenario provided in the initial issue -- an empty include directive -- may be about all that's actually handled by this change. Adding an include to a non-existent script (properly) throws a `FileNotFoundException`. Adding an include to a script with syntax errors in it ignores the syntax errors entirely, I guess because all the parser is doing is traversing files and concatenating them.

I would love to be able to do more than this, particularly identifying syntax errors in pre- and post-deployment scripts, but that will require significantly more work and I'm not certain whether it's doable or not. My recommendation would be to merge this as it resolves the specific issue, and then I'd be open to adding a new issue to see if we can improve upon this further.

**Review by jmezach — APPROVED** [review]
@jeffrosenberg This looks good to me. I agree that it would be useful to have meaningful parser errors for included scripts but as you said it might be tricky to pull off. Perhaps we can do something with the code we're including from Microsoft, but we can always figure that out later. For now, let's get this merged and released :).

**Pull request merged**